# Exploring E-bay Car Sales Data
- This Data is taken from Kaggle Competitions

### Data Dictionary

> <font color=blue>**_dateCrawled_**</font> - When this ad was first crawled. All field-values are taken from this date.<br>
> <font color=blue>**_name_**</font> - Name of the car.<br>
> <font color=blue>**_seller_**</font> - Whether the seller is private or a dealer.<br>
> <font color=blue>**_offerType_**</font> - The type of listing<br>
> <font color=blue>**_price_**</font> - The price on the ad to sell the car.<br>
> <font color=blue>**_abtest_**</font> - Whether the listing is included in an A/B test.<br>
> <font color=blue>**_vehicleType_**</font> - The vehicle Type.<br>
> <font color=blue>**_yearOfRegistration_**</font> - The year in which which year the car was first registered.<br>
> <font color=blue>**_gearbox_**</font> - The transmission type.<br>
> <font color=blue>**_powerPS_**</font> - The power of the car in PS.<br>
> <font color=blue>**_model_**</font> - The car model name.<br>
> <font color=blue>**_kilometer_**</font> - How many kilometers the car has driven.<br>
> <font color=blue>**_monthOfRegistration_**</font> - The month in which which year the car was first registered.<br>
> <font color=blue>**_fuelType_**</font> - What type of fuel the car uses.
> <font color=blue>**_brand_**</font> - The brand of the car.<br>
> <font color=blue>**_notRepairedDamage_**</font> - If the car has a damage which is not yet repaired.<br>
> <font color=blue>**_dateCreated_**</font> - The date on which the eBay listing was created.<br>
> <font color=blue>**_nrOfPictures_**</font> - The number of pictures in the ad.<br>
> <font color=blue>**_postalCode_**</font> - The postal code for the location of the vehicle.<br>
> <font color=blue>**_lastSeenOnline_**</font> - When the crawler saw this ad last online.<br>

### <font color=Orange>_Aim_</font>
We aim to clean the data and analyze the included used car listings


In [1]:
import pandas as pd
import numpy as np
autos = pd.read_csv('autos.csv', encoding="Latin-1")

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

### _Key Observations:_

- The Dataset includes of 20 Columns, most of which are Strings

- Some Columns have null values, but none have more than 20% null values.

In [5]:
column_names = autos.columns
column_names

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
import re

def clean_col(col):
    col.strip()
    col = col.replace("yearOfRegistration",
                      "registration_year")
    col = col.replace("monthOfRegistration", 
                      "registration_month")
    col = col.replace("notRepairedDamage", 
                      "unrepaired_damage")
    col = col.replace("dateCreated", 
                      "ad_created")
    return re.sub('([a-z0-9])([A-Z])', r'\1_\2',col).lower()
    
autos.columns = [clean_col(c) for c in autos.columns]
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [7]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

***Changed Names for all the columnms from camelCase to snake_case. For e.g. 'nrOfPictures' became 'nr_of_pictures'***

In [8]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-19 17:36:18  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

### _Key Observations:_
- **nr_of_pictures**, **seller** and **offer_type** columns have mostly a single value. And they should be dropped.
- Columns **price** and **odometer** are shown as text (Object) type, whereas they should be numeric (int) type and should be converted.

In [9]:
autos["price"] = autos["price"].replace(
    {'\$':'',',':''}, 
    regex=True).astype(int)
autos["price"]

0         5000
1         8500
2         8990
3         4350
4         1350
5         7900
6          300
7         1990
8          250
9          590
10         999
11         350
12        5299
13        1350
14        3999
15       18900
16         350
17        5500
18         300
19        4150
20        3500
21       41500
22       25450
23        7999
24       48500
25          90
26         777
27           0
28        5250
29        4999
         ...  
49970    15800
49971      950
49972     3300
49973     6000
49974        0
49975     9700
49976     5900
49977     5500
49978      900
49979    11000
49980      400
49981     2000
49982     1950
49983      600
49984        0
49985     1000
49986    15900
49987    21990
49988     9550
49989      150
49990    17500
49991      500
49992     4800
49993     1650
49994     5000
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 50000, dtype: int64

In [10]:
autos["odometer"] = autos["odometer"].replace(
    {'km':'',',':''},
    regex=True).astype(int)
autos.rename(columns={'odometer':'odometer_km'}, 
             inplace=True)
autos['odometer_km']

0        150000
1        150000
2         70000
3         70000
4        150000
5        150000
6        150000
7        150000
8        150000
9        150000
10       150000
11       150000
12        50000
13       150000
14       150000
15        80000
16       150000
17       150000
18       150000
19       150000
20       150000
21       150000
22        10000
23       150000
24        30000
25       150000
26       125000
27       150000
28       150000
29       150000
          ...  
49970     60000
49971    150000
49972    150000
49973    150000
49974    150000
49975    100000
49976    150000
49977    150000
49978    150000
49979     70000
49980    125000
49981    150000
49982     90000
49983    150000
49984    150000
49985    150000
49986    125000
49987     50000
49988    150000
49989    150000
49990     30000
49991    150000
49992    125000
49993    150000
49994    150000
49995    100000
49996    150000
49997      5000
49998     40000
49999    150000
Name: odometer_km, Lengt

In [11]:
autos.drop(columns=
           ['nr_of_pictures','seller','offer_type'], inplace=True)
autos.describe(include='all')

date_crawled         name         price abtest vehicle_type  \
count                 50000        50000  5.000000e+04  50000        44905   
unique                48213        38754           NaN      2            8   
top     2016-03-19 17:36:18  Ford_Fiesta           NaN   test    limousine   
freq                      3           78           NaN  25756        12859   
mean                    NaN          NaN  9.840044e+03    NaN          NaN   
std                     NaN          NaN  4.811044e+05    NaN          NaN   
min                     NaN          NaN  0.000000e+00    NaN          NaN   
25%                     NaN          NaN  1.100000e+03    NaN          NaN   
50%                     NaN          NaN  2.950000e+03    NaN          NaN   
75%                     NaN          NaN  7.200000e+03    NaN          NaN   
max                     NaN          NaN  1.000000e+08    NaN          NaN   

        registration_year  gearbox      power_ps  model    odometer_km  \
count        50000.000000    47320  50000.000000  47242   50000.000000   
unique                NaN        2           NaN    245            NaN   
top                   NaN  manuell           NaN   golf            NaN   
freq                  NaN    36993           NaN   4024            NaN   
mean          2005.073280      NaN    116.355920    NaN  125732.700000   
std            105.712813      NaN    209.216627    NaN   40042.211706   
min           1000.000000      NaN      0.000000    NaN    5000.000000   
25%           1999.000000      NaN     70.000000    NaN  125000.000000   
50%           2003.000000      NaN    105.000000    NaN  150000.000000   
75%           2008.000000      NaN    150.000000    NaN  150000.000000   
max           9999.000000      NaN  17700.000000    NaN  150000.000000   

        registration_month fuel_type       brand unrepaired_damage  \
count         50000.000000     45518       50000             40171   
unique                 NaN         7          40                 2   
top                    NaN    benzin  volkswagen              nein   
freq                   NaN     30107       10687             35232   
mean              5.723360       NaN         NaN               NaN   
std               3.711984       NaN         NaN               NaN   
min               0.000000       NaN         NaN               NaN   
25%               3.000000       NaN         NaN               NaN   
50%               6.000000       NaN         NaN               NaN   
75%               9.000000       NaN         NaN               NaN   
max              12.000000       NaN         NaN               NaN   

                 ad_created   postal_code            last_seen  
count                 50000  50000.000000                50000  
unique                   76           NaN                39481  
top     2016-04-03 00:00:00           NaN  2016-04-07 06:17:27  
freq                   1946           NaN                    8  
mean                    NaN  50813.627300                  NaN  
std                     NaN  25779.747957                  NaN  
min                     NaN   1067.000000                  NaN  
25%                     NaN  30451.000000                  NaN  
50%                     NaN  49577.000000                  NaN  
75%                     NaN  71540.000000                  NaN  
max                     NaN  99998.000000                  NaN

In [12]:
autos['price'].value_counts().sort_index()

0           1421
1            156
2              3
3              1
5              2
8              1
9              1
10             7
11             2
12             3
13             2
14             1
15             2
17             3
18             1
20             4
25             5
29             1
30             7
35             1
40             6
45             4
47             1
49             4
50            49
55             2
59             1
60             9
65             5
66             1
            ... 
151990         1
155000         1
163500         1
163991         1
169000         1
169999         1
175000         1
180000         1
190000         1
194000         1
197000         1
198000         1
220000         1
250000         1
259000         1
265000         1
295000         1
299000         1
345000         1
350000         1
999990         1
999999         2
1234566        1
1300000        1
3890000        1
10000000       1
11111111       2
12345678      

In [13]:
autos = autos[autos['price'].between(100,1000000)]
autos['price'].value_counts().sort_index()

100       134
110         3
111         2
115         2
117         1
120        39
122         1
125         8
129         1
130        15
135         1
139         1
140         9
145         2
149         7
150       224
156         2
160         8
170         7
173         1
175        12
179         1
180        35
185         1
188         1
190        16
193         1
195         2
198         1
199        41
         ... 
120000      2
128000      1
129000      1
130000      1
135000      1
137999      1
139997      1
145000      1
151990      1
155000      1
163500      1
163991      1
169000      1
169999      1
175000      1
180000      1
190000      1
194000      1
197000      1
198000      1
220000      1
250000      1
259000      1
265000      1
295000      1
299000      1
345000      1
350000      1
999990      1
999999      2
Name: price, Length: 2312, dtype: int64

***Observed outliers (below 100 or above 1000000), in 'price' column and removed them.***

In [14]:
autos['odometer_km'].value_counts().sort_index()

5000        760
10000       245
20000       757
30000       777
40000       814
50000      1009
60000      1153
70000      1214
80000      1412
90000      1733
100000     2101
125000     5038
150000    31214
Name: odometer_km, dtype: int64

In [15]:
autos = autos[autos['registration_month']!=0]

In [16]:
autos['registration_month'].value_counts()

3     4981
6     4255
4     4020
5     4016
7     3842
10    3577
12    3359
9     3318
11    3305
1     3199
8     3115
2     2927
Name: registration_month, dtype: int64

In [17]:
autos['registration_year'].value_counts().sort_index()

1800       2
1927       1
1929       1
1931       1
1934       1
1937       3
1938       1
1939       1
1941       2
1943       1
1948       1
1950       1
1951       2
1952       1
1953       1
1954       1
1955       2
1956       4
1957       1
1958       2
1959       6
1960      13
1961       6
1962       4
1963       6
1964      11
1965      16
1966      20
1967      22
1968      23
        ... 
1995     965
1996    1149
1997    1683
1998    2059
1999    2601
2000    2572
2001    2410
2002    2315
2003    2533
2004    2585
2005    2690
2006    2576
2007    2193
2008    2137
2009    2033
2010    1560
2011    1594
2012    1291
2013     792
2014     654
2015     368
2016     911
2017    1051
2018     365
2019       2
2800       1
4800       1
5000       1
6200       1
9000       1
Name: registration_year, Length: 86, dtype: int64

In [18]:
autos = autos[autos['registration_year'].between(1900,2017)]
autos['registration_year'].value_counts().sort_index()

1927       1
1929       1
1931       1
1934       1
1937       3
1938       1
1939       1
1941       2
1943       1
1948       1
1950       1
1951       2
1952       1
1953       1
1954       1
1955       2
1956       4
1957       1
1958       2
1959       6
1960      13
1961       6
1962       4
1963       6
1964      11
1965      16
1966      20
1967      22
1968      23
1969      16
        ... 
1988     119
1989     147
1990     247
1991     298
1992     317
1993     351
1994     532
1995     965
1996    1149
1997    1683
1998    2059
1999    2601
2000    2572
2001    2410
2002    2315
2003    2533
2004    2585
2005    2690
2006    2576
2007    2193
2008    2137
2009    2033
2010    1560
2011    1594
2012    1291
2013     792
2014     654
2015     368
2016     911
2017    1051
Name: registration_year, Length: 78, dtype: int64

***Observed outliers for 'Registration month' and 'Registration year'.***<br>
***For 'Registration month' it was value of 0 and 'Registration Year' it was below 1900 and above 2017<br>Removed those outliers and cleaned the Registration Month and Registration Year Column.***

In [19]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0   5000  control          bus               2004    manuell       158   
1   8500  control    limousine               1997  automatik       286   
2   8990     test    limousine               2009    manuell       102   
3   4350  control   kleinwagen               2007  automatik        71   
4   1350     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

In [20]:
autos['date_crawled'] = autos['date_crawled'].str[:10]
autos['date_crawled'] = autos['date_crawled'].str.replace('-','').astype(int)

autos['last_seen'] = autos['last_seen'].str[:10]
autos['last_seen'] = autos['last_seen'].str.replace('-','').astype(int)

autos['ad_created'] = autos['ad_created'].str[:10]
autos['ad_created'] = autos['ad_created'].str.replace('-','').astype(int)

In [21]:
autos.head()

date_crawled                                               name  price  \
0      20160326                   Peugeot_807_160_NAVTECH_ON_BOARD   5000   
1      20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   8500   
2      20160326                         Volkswagen_Golf_1.6_United   8990   
3      20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   4350   
4      20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   1350   

    abtest vehicle_type  registration_year    gearbox  power_ps   model  \
0  control          bus               2004    manuell       158  andere   
1  control    limousine               1997  automatik       286     7er   
2     test    limousine               2009    manuell       102    golf   
3  control   kleinwagen               2007  automatik        71  fortwo   
4     test        kombi               2003    manuell         0   focus   

   odometer_km  registration_month fuel_type       brand unrepaired_damage  \
0       150000                   3       lpg     peugeot              nein   
1       150000                   6    benzin         bmw              nein   
2        70000                   7    benzin  volkswagen              nein   
3        70000                   6    benzin       smart              nein   
4       150000                   7    benzin        ford              nein   

   ad_created  postal_code  last_seen  
0    20160326        79588   20160406  
1    20160404        71034   20160406  
2    20160326        35394   20160406  
3    20160312        33729   20160315  
4    20160401        39218   20160401

In [22]:
autos['date_crawled'].value_counts(normalize=True, dropna=False).sort_index()

20160305    0.025287
20160306    0.014286
20160307    0.036311
20160308    0.033050
20160309    0.032889
20160310    0.032361
20160311    0.033119
20160312    0.036932
20160313    0.015526
20160314    0.036656
20160315    0.033877
20160316    0.029444
20160317    0.031328
20160318    0.012862
20160319    0.034428
20160320    0.038034
20160321    0.036932
20160322    0.032797
20160323    0.032384
20160324    0.029169
20160325    0.031121
20160326    0.032361
20160327    0.031190
20160328    0.035140
20160329    0.033854
20160330    0.033486
20160331    0.031672
20160401    0.034244
20160402    0.035783
20160403    0.039045
20160404    0.036633
20160405    0.013160
20160406    0.003215
20160407    0.001424
Name: date_crawled, dtype: float64

***DATE CRAWLED - The above distribbution look like a uniform ditribution.<br>Which means that the ads were crawled on a regular basis, with an average of 3% ads being crawled daily for 34 days.<br>***

In [23]:
autos['ad_created'].value_counts(normalize=True, dropna=False).sort_index()

20150611    0.000023
20150810    0.000023
20150909    0.000023
20151110    0.000023
20151205    0.000023
20151230    0.000023
20160103    0.000023
20160107    0.000023
20160110    0.000046
20160113    0.000023
20160114    0.000023
20160116    0.000023
20160122    0.000023
20160127    0.000069
20160129    0.000023
20160201    0.000023
20160202    0.000046
20160205    0.000046
20160207    0.000023
20160208    0.000023
20160209    0.000023
20160211    0.000023
20160212    0.000046
20160214    0.000046
20160216    0.000023
20160217    0.000023
20160218    0.000046
20160219    0.000069
20160220    0.000046
20160221    0.000069
              ...   
20160309    0.033004
20160310    0.032085
20160311    0.033372
20160312    0.036771
20160313    0.016789
20160314    0.035370
20160315    0.033647
20160316    0.029927
20160317    0.031029
20160318    0.013574
20160319    0.033280
20160320    0.037988
20160321    0.037207
20160322    0.032614
20160323    0.032223
20160324    0.029100
20160325    0

***AD CREATED - The above distribbution look like a left skewed ditribution.Which means that most of the ads were created recently.<br>It could be beacuse the cars that were posted earlier (like a month or two ago) would have got sold, hence their ad was pulled down. <br>Leaving only a very small number of ads (of unsold cars) posted from a long period of time.***

In [24]:
autos['last_seen'].value_counts(normalize=True, dropna=False).sort_index()

20160305    0.001102
20160306    0.004134
20160307    0.005466
20160308    0.007212
20160309    0.009577
20160310    0.010611
20160311    0.012379
20160312    0.023610
20160313    0.008705
20160314    0.012425
20160315    0.015733
20160316    0.016284
20160317    0.027584
20160318    0.007350
20160319    0.015618
20160320    0.020280
20160321    0.020464
20160322    0.020877
20160323    0.018420
20160324    0.019568
20160325    0.018833
20160326    0.016514
20160327    0.015181
20160328    0.020648
20160329    0.022209
20160330    0.024483
20160331    0.023565
20160401    0.023220
20160402    0.025057
20160403    0.025057
20160404    0.024024
20160405    0.126941
20160406    0.223473
20160407    0.133395
Name: last_seen, dtype: float64

***LAST SEEN - Here the distribution is heavily left skewed, with almost 50% of the ads being seen in last 3 days of the dataset.<br>We can give a similar analogy to 'ad_created', older car ads are in less quantity in last seen, as they might be sold already.<br>And, as a consequence most of the last seens ads are of recent sellers, who are interested in selling their cars ASAP.***



In [25]:
autos['registration_year'].describe()

count    43540.000000
mean      2003.543753
std          7.166670
min       1927.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       2017.000000
Name: registration_year, dtype: float64

In [26]:
autos['registration_year'].value_counts(normalize=True, dropna=False).sort_index()

1927    0.000023
1929    0.000023
1931    0.000023
1934    0.000023
1937    0.000069
1938    0.000023
1939    0.000023
1941    0.000046
1943    0.000023
1948    0.000023
1950    0.000023
1951    0.000046
1952    0.000023
1953    0.000023
1954    0.000023
1955    0.000046
1956    0.000092
1957    0.000023
1958    0.000046
1959    0.000138
1960    0.000299
1961    0.000138
1962    0.000092
1963    0.000138
1964    0.000253
1965    0.000367
1966    0.000459
1967    0.000505
1968    0.000528
1969    0.000367
          ...   
1988    0.002733
1989    0.003376
1990    0.005673
1991    0.006844
1992    0.007281
1993    0.008062
1994    0.012219
1995    0.022164
1996    0.026390
1997    0.038654
1998    0.047290
1999    0.059738
2000    0.059072
2001    0.055351
2002    0.053169
2003    0.058176
2004    0.059371
2005    0.061782
2006    0.059164
2007    0.050367
2008    0.049081
2009    0.046693
2010    0.035829
2011    0.036610
2012    0.029651
2013    0.018190
2014    0.015021
2015    0.0084

***REGISTRATION YEAR - We can observe that the distribution is left skewed.<br>We can observe that most of the cars are registered between 1998-2008.<br>Since, most of the people use their cars for 10-15 years and then sell/discard them.***

In [27]:
brands = autos['brand'].value_counts(normalize=True)
results = brands >= 0.05

# Finding names of brands wit more then 5% share of ads
top_brands_name = brands[results].index
print(top_brands_name)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [29]:
brand_mean_price = {}

for car in top_brands_name:
    mean_price = autos.loc[autos['brand']== car,'price'].mean()
    brand_mean_price[car] = round(mean_price, 2)
    
brand_mean_price

{'audi': 9718.9,
 'bmw': 8637.41,
 'ford': 4289.74,
 'mercedes_benz': 8846.68,
 'opel': 3173.85,
 'volkswagen': 5943.71}

***We found Automobile Brands with more then 5% share of the Ads.<br>And from those names we derived the average price of those top brands vehicles.<br>Those average prices are as follows:***

> **Volkswagen:** 5943.71<br>
> **BMW:** 8637.41<br>
> **Opel:** 3173.85<br>
> **Mercedes Benz:** 8846.68<br>
> **Audi:** 9718.9<br>
> **Ford:** 4289.74<br>


***Here we may notice that the most expensive vehicles are from Audi, BMW and Mercedes.<br>While Ford, Opel are the most economical options, Volkswagen is in between***

In [30]:
brand_mean_mileage = {}

for car in top_brands_name:
    mean_mileage = autos.loc[autos['brand'] == car, 'odometer_km'].mean()
    brand_mean_mileage[car] = round(mean_mileage, 2)

brand_mean_mileage

{'audi': 128830.76,
 'bmw': 132575.41,
 'ford': 123839.23,
 'mercedes_benz': 131034.56,
 'opel': 128805.64,
 'volkswagen': 128365.67}

In [31]:
bmp_series = pd.Series(brand_mean_price).sort_values(ascending=False)
bmm_series = pd.Series(brand_mean_mileage).sort_values(ascending=False)
print(bmp_series)
bmm_series

audi             9718.90
mercedes_benz    8846.68
bmw              8637.41
volkswagen       5943.71
ford             4289.74
opel             3173.85
dtype: float64


bmw              132575.41
mercedes_benz    131034.56
audi             128830.76
opel             128805.64
volkswagen       128365.67
ford             123839.23
dtype: float64

In [32]:
top_brands_df = pd.DataFrame(bmp_series, columns=['mean_price'])
top_brands_df

mean_price
audi              9718.90
mercedes_benz     8846.68
bmw               8637.41
volkswagen        5943.71
ford              4289.74
opel              3173.85

In [33]:
top_brands_df['mean_mileage'] = bmm_series
top_brands_df

mean_price  mean_mileage
audi              9718.90     128830.76
mercedes_benz     8846.68     131034.56
bmw               8637.41     132575.41
volkswagen        5943.71     128365.67
ford              4289.74     123839.23
opel              3173.85     128805.64

***Here we can observe that car mileage/brand does not vary much as compared to the car prices/brand.<br><hr>While mean price difference for most expensive car brand (Audi) and most economical car brand (Opel) was a gigantic 71%.<br>However, same is not true for mean mileage, with total difference between top mileage giving brand(BMW) & least mileage giving brand(Ford) being mere 7%.***